keps/sig-node/2535-ensure-secret-pulled-images/README.md

```diff
@@ -165,57 +201,367 @@ ensure all tenants have rights to the images that are already present on a host.
+      pull attempts.
+
+The kubelet uses the following Golang interface to interact with the cache:
+```go
+// ImagePullManager keeps the state of images that were pulled and which are
+// currently still being pulled.
+// It should keep an internal state of images currently being pulled by the kubelet
+// in order to determine whether to destroy a "pulling" record should an image
+// pull fail.
+type ImagePullManager interface {
+    // RecordPullIntent records an intent to pull an image and should be called
+    // before a pull of the image occurs.
+    //
+    // RecordPullIntent() should be called before every image pull. Each call of
+    // RecordPullIntent() must match exactly one call of RecordImagePulled()/RecordImagePullFailed().
+    //
+    // `image` is the content of the pod's container `image` field.
+    RecordPullIntent(image string) error
+    // RecordImagePulled writes a record of an image being successfully pulled
+    // with ImagePullCredentials.
+    //
+    // If `credentials` is nil, it means the image was pulled successfully, but
+    // a reverification should be required upon next pull request. This can be
+    // useful in cases where the kubelet crashed in the middle of the pull but
+    // the image appears on the node after the kubelet restarts.
+    //
+    // `image` is the content of the pod's container `image` field.
+    RecordImagePulled(image, imageRef string, credentials *imagemanager.ImagePullCredentials) error
+    // RecordImagePullFailed should be called if an image failed to pull.
+    //
+    // Internally, it lowers its reference counter for the given image. If the
+    // counter reaches zero, the pull intent record for the image is removed.
+    //
+    // `image` is the content of the pod's container `image` field.
+    RecordImagePullFailed(image string) error
+    // ShouldReverifyCredentials evaluates the policy for image specified in
+    // "image" and if the policy demands reverification, it checks the internal
+    // cache to see if there's a record of pulling the image with the presented
+    // set of credentials or if the image can be accessed by any of the node's pods.
+    //
+    // Returns true if the policy demands reverification and no record of the pull
+    // was found in the cache.
+    //
+    // `image` is the content of the pod's container `image` field.
+    ShouldReverifyCredentials(image, imageRef string, credentials ... *imagemanager.ImagePullSecret) bool
+    // DestroyImagePulledRecord removes an ImagePulledRecord from the kubelet's
+    // storage.
+    // WARNING!
+    //     This method is purposed for the node image GC to be able to remove an
+    //     ImagePulledRecord of an image it deleted from the node.
+    //     Using this method outside the GC context might cause security issues
+    //     if the image represented by the `imageRef` still exists on the node at
+    //     the time this method is called.
```

Comment:
If we go with the `remove(imageRef)` approach, document this should be called *after* the image is actually removed successfully.

An alternative is to pass in all known image refs (which the kubelet / gc periodically list anyway) and let the implementation clean up pulled records that aren't in that list. That would self-heal the leaked record case eventually, but may introduce races between image pulls and image listing.

```
GC thread:
  T0: token = intenttoprune() // token could be a timestamp or a generation, etc
  T10: knownImages = list all images
  T20: maybe gc stuff
  T30: pruneunknownrecords(knownImages, token) // prunes to knownImages+things recorded since intenttoprune was called (based on token)

kubelet pull thread(s)
  T15: kubelet records intent to pull foo
  T25: kubelet successfully pulls foo and records success (in a way that pruneunknownrecords can fold in)
```

I think the paired IntentToPrune / PruneUnknownRecords approach where the full list of images known to GC is passed in works better.